pkg/svcat/service-catalog/plan.go

```diff
@@ -102,56 +128,82 @@ func (sdk *SDK) RetrievePlans(opts RetrievePlanOptions) ([]Plan, error) {
+
+	listOpts := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(FieldExternalPlanName, name).String(),
 	}
-	return &searchResults.Items[0], nil
+
+	return sdk.retrieveSinglePlanByListOptions(name, opts, listOpts)
 }
 
-// RetrievePlanByID gets a plan by its UUID.
-func (sdk *SDK) RetrievePlanByID(uuid string) (*v1beta1.ClusterServicePlan, error) {
-	plan, err := sdk.ServiceCatalog().ClusterServicePlans().Get(uuid, v1.GetOptions{})
-	if err != nil {
-		return nil, fmt.Errorf("unable to get plan by uuid '%s' (%s)", uuid, err)
+// RetrievePlanByClassAndName gets a plan by its external name and class name combination.
+func (sdk *SDK) RetrievePlanByClassAndName(className, planName string, opts ScopeOptions) (Plan, error) {
+	if opts.Scope == AllScope {
+		return nil, errors.New("invalid scope: all")
 	}
-	return plan, nil
-}
 
-// RetrievePlanByClassAndPlanNames gets a plan by its class/plan name combination.
-func (sdk *SDK) RetrievePlanByClassAndPlanNames(className, planName string,
-) (*v1beta1.ClusterServicePlan, error) {
-	class, err := sdk.RetrieveClassByName(className, ScopeOptions{Scope: ClusterScope})
+	class, err := sdk.RetrieveClassByName(className, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	planOpts := v1.ListOptions{
+	var classRefSelector fields.Selector
+	if opts.Scope.Matches(ClusterScope) {
+		classRefSelector = fields.OneTermEqualSelector(FieldClusterServiceClassRef, class.GetName())
+	} else {
+		classRefSelector = fields.OneTermEqualSelector(FieldServiceClassRef, class.GetName())
+	}
+
+	listOpts := metav1.ListOptions{
 		FieldSelector: fields.AndSelectors(
-			fields.OneTermEqualSelector(FieldServiceClassRef, class.GetName()),
+			classRefSelector,
 			fields.OneTermEqualSelector(FieldExternalPlanName, planName),
 		).String(),
 	}
-	searchResults, err := sdk.ServiceCatalog().ClusterServicePlans().List(planOpts)
+
+	ss := []string{class.GetName(), planName}
+	return sdk.retrieveSinglePlanByListOptions(strings.Join(ss, "/"), opts, listOpts)
+}
+
+func (sdk *SDK) retrieveSinglePlanByListOptions(name string, scopeOpts ScopeOptions, listOpts metav1.ListOptions) (Plan, error) {
+	plans, err := sdk.retrievePlansByListOptions(scopeOpts, listOpts)
 	if err != nil {
-		return nil, fmt.Errorf("unable to search plans by class/plan name '%s/%s' (%s)", className, planName, err)
+		return nil, err
 	}
-	if len(searchResults.Items) == 0 {
-		return nil, fmt.Errorf("plan not found '%s/%s'", className, planName)
+	if len(plans) == 0 {
+		return nil, fmt.Errorf("plan not found '%s'", name)
 	}
-	if len(searchResults.Items) > 1 {
-		// Note: Should never occur, as class/plan name combo must be unique
-		return nil, fmt.Errorf("more than one matching plan found for '%s/%s'", className, planName)
+	if len(plans) > 1 {
+		return nil, fmt.Errorf("more than one matching plan found for '%s'", name)
 	}
-	return &searchResults.Items[0], nil
+	return plans[0], nil
+}
+
+// RetrievePlanByID gets a plan by its UUID.
```

Comment:
FYI to other reviewers, we have an open issue to fix how svcat names things (k8s name = uuid in this context unfortunately). See https://github.com/kubernetes-incubator/service-catalog/issues/2397

Reply:
🤦‍♀️ Heh, you are being too nice. That was a mistake on my part, you are correct that validation doesn't belong in this function! 😀